Asset Store/Update/Destroy Actions

spencerrlongg opened the pull request:
# Description
This PR introduces the concept of actions to both Assets controllers to begin to unify the way we do things, as well as replacing a lot of the custom logic that the BulkAssetsController's update method performed. 

Now, instead of disparate methods for updating one or many assets, we perform the same logic either way and loop over the action to do it multiple times. The business logic is largely removed from the controllers in these situations, and the logic that is performed inside them consists of setting up the arrays, looping, and catching/returning errors or success. 

At first glance the long method signatures can seem unwieldy, but I've come around to them and kind of like how specific they are, by doing this you can use the `StoreAssetAction` how ever you'd like. 

For example: 
The `StoreAssetAction` only _really_ requires three arguments, `model_id`, `status_id`, and at least for now an instance of `ImageUploadRequest` (I'll get into why that is later). Everything else is optional. 

Using named parameters (PHP 8.1 and up) this makes the action very flexible, and we can use it anywhere we'd like to make a new asset, including on the CLI. PHPStorm also takes advantage of named parameters and method signatures very well, making it hard to get the action wrong in practice.
`StoreAssetAction::run(model_id: $request->validated('model_id'), status_id: $request->validated('status_id'), $request, ...)`

**Why are we passing the full request in?**
1. The `handleImages()` method on the `ImageUploadRequest` is what does all the image processing, and it uses a lot of request context to do so at the moment. I'd like to see this moved out into some kind of helper method or maybe promoted up to the extended request (as in StoreAssetRequest) - but that seemed like a bit much for this PR. 

2. Custom fields logic is also based on the request currently because the code doesn't necessarily know what the possible custom fields in the request are going to be. This is also solvable, but will wait on that for a little bit. 

In general, we don't want to pass the request beyond the controller - requests should stay at the Http layer and actions should be able to function purely, without any knowledge of the request, but again because we require it for handling images we're willing to break some rules in the hopes that it will change in the future. 

Also, the pattern of try/catching all Actions ensures that eventually we should be able to get to a place where a user should never encounter a 500, the `report()` method ensures that we are still alerted by them though. 

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Existing tests pass, and some new ones written. 

**Test Configuration**:
* PHP version: 8.1
* MySQL version 8.1

spencerrlongg committed:
just a little scaffolding

spencerrlongg committed:
bs

spencerrlongg committed:
alright, this mostly works.

spencerrlongg committed:
almost wrapped up StoreAssetAction.php

spencerrlongg committed:
no more base action, scaffolded update and delete

spencerrlongg committed:
delete done

spencerrlongg committed:
a couple more tests passing

spencerrlongg committed:
a few fixes, tests pass

spencerrlongg committed:
good progress on update action

spencerrlongg committed:
some more good progress on UpdateAssetAction.php

spencerrlongg committed:
gui seemingly working, other good progress - api next

spencerrlongg committed:
var changes

spencerrlongg committed:
fixed some things, made bulk progress, tests pass

spencerrlongg committed:
cleaned - more work needs to happen

spencerrlongg committed:
rm extra code

spencerrlongg committed:
tests 'n stuff

spencerrlongg committed:
Add CustomFieldPermissionException handling

spencerrlongg committed:
Fix error handling in BulkAssetsController

Updated error handling to ensure all exceptions are properly stored in the errors array and removed unused custom_field_problem variable. Also, uncommented location assignment code in UpdateAssetAction for clearer logic.

spencerrlongg committed:
Fix bulk asset update error handling and improve access checks

Improved the error handling in the BulkAssetsController by re-enabling custom field permission checks. Updated the UpdateAssetAction to better handle encrypted custom fields and streamlined user roles in the BulkEditAssetsTest.

spencerrlongg committed:
Allow null values for asset dates in update action

This commit improves the asset update functionality by allowing null values for purchase date, next audit date, and expected check-in date when specified in the request. Additionally, minor corrections were made in the test code to enhance readability and correctness.

spencerrlongg committed:
Update asset model association and location logic handling

Refactor the asset update process to properly associate models and handle location ID changes based on request input. Introduce checks for model_id, company_id, and last_audit_date to ensure correct updates. Add commentary to clarify the handling of location ID logic and temporarily resolve issues with location assignment when requests do not expect JSON responses.

spencerrlongg committed:
Fix typos and use correct array notation in asset updates

Corrected spelling mistakes in test files and updated code to use the correct array notation in request validation for asset updates. This ensures better readability and consistency in the codebase while preventing potential errors related to incorrect data handling.

spencerrlongg committed:
Handle exceptions when saving assets in AssetImporter

Added try-catch blocks to manage ValidationException and generic exceptions during asset saving. This ensures proper logging of errors and reporting of unexpected issues, improving robustness and error handling in the asset import process.

spencerrlongg committed:
Fix error handling in BulkAssetsController

Updated exception handling to capture detailed validation errors in BulkAssetsController. Corrected formatting in notifications view for displaying error messages.

spencerrlongg committed:
Fix error message format in BulkAssetsController

Previously, the error message was set as a string, which led to an inconsistency in how errors were handled. This change updates the error message to be an array, ensuring it aligns with the expected format.

spencerrlongg committed:
get rid of comment

spencerrlongg committed:
Remove debug dump statements

This commit removes various debug dump statements from BulkAssetsController and UpdateAssetAction. These changes clean up the code and improve readability by eliminating unnecessary debugging output. Maintaining clean code ensures better maintainability and performance.

spencerrlongg committed:
Remove unused code from UpdateAssetAction.php

Removed commented-out code related to handling encrypted custom fields in UpdateAssetAction.php. This cleanup helps to maintain a more readable and maintainable codebase.

spencerrlongg committed:
Refactor error handling and remove redundant code

Refactored error handling by replacing `\Exception` with `Exception` for consistency. Removed redundant comments and unused code to improve readability and maintainability of the `StoreAssetAction` class.

spencerrlongg committed:
Add CheckoutNotAllowed exception in UpdateAssetAction

This change introduces the CheckoutNotAllowed exception to the UpdateAssetAction class. It ensures that the action can handle scenarios where checking out an asset is not permitted, improving the robustness of the application.

spencerrlongg committed:
Refactor UpdateAssetAction for clarity and remove comments

Removed outdated and redundant comments for improved code clarity and readability. Renamed `bulkUpdate` method to `bulkLocationUpdate` for better context and understanding.

spencerrlongg committed:
Add tests for storing assets with all fields and multiple entries

Introduced StoreAssetTest to verify the handling of full asset records and multiple asset entries in the database. Updated StoreAssetAction and AssetsController to support next audit date validation and improved error handling.

spencerrlongg committed:
Remove unnecessary parameter and add delete asset tests

Removed the redundant $request parameter from the destroy method in AssetsController. Added feature tests to verify asset deletion functionality with proper permissions and ensure deletion is forbidden without permissions.

spencerrlongg committed:
Add permissions checks to asset tests

Implement tests to ensure users without appropriate permissions are denied access to asset endpoints. Update tests to verify proper soft deletions and asset existence checks in delete scenarios.

spencerrlongg committed:
Fix authorization check in AssetsController

Removed incorrect and commented-out authorization check in the destroy method. Ensured proper authorization by explicitly authorizing the asset instance before attempting deletion.

spencerrlongg commented:
changed to draft because of all the changes I wasn't aware of in the controllers

spencerrlongg committed:
a little work for conflicts

spencerrlongg committed:
Simplify asset handling logic for creation and update

Refactor asset creation and update processes by removing redundant save checks and organizing error handling logic. Enhance success and failure feedback in the asset creation flow to improve clarity and user experience. Add support for detailed multi-success and partial failure messages to alert users of outcomes.

spencerrlongg committed:
Merge remote-tracking branch 'upstream/develop' into refactor/asset-action

# Conflicts:
#	app/Http/Controllers/Api/AssetsController.php
#	app/Http/Controllers/Assets/AssetsController.php
#	app/Http/Controllers/Assets/BulkAssetsController.php
#	resources/lang/en-US/admin/hardware/message.php

spencerrlongg committed:
Remove unused imports from AssetsController files.

Simplify the code by cleaning up unused import statements in both `AssetsController.php` and `Api/AssetsController.php`. This reduces potential confusion and helps maintain cleaner code for future development and maintenance.

spencerrlongg committed:
formatting

marcusmoore commented on app/Actions/Assets/DestroyAssetAction.php:
```diff
@@ -0,0 +1,38 @@
+        if ($asset->image) {
+            try {
+                Storage::disk('public')->delete('assets'.'/'.$asset->image);
+            } catch (\Exception $e) {
```
Let's import this por favor

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
```diff
@@ -0,0 +1,143 @@
+use Illuminate\Support\Facades\Crypt;
+use Illuminate\Support\Facades\Gate;
+use Illuminate\Support\Facades\Log;
+use Illuminate\Support\MessageBag;
```
These two can be removed since they're not used.

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
```diff
@@ -0,0 +1,143 @@
+        $asset_tag = null,
+        $order_number = null,
+        $notes = null,
+        $user_id = null,
```
`$user_id` isn't used and can be removed I think.

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
```diff
@@ -0,0 +1,143 @@
+        $requestable = null,
+        $rtd_location_id = null,
+        $location_id = null, //do something with this
+        $files = null,
```
I think `$files` can be removed

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
```diff
@@ -0,0 +1,143 @@
+        $supplier_id = null,
+        $requestable = null,
+        $rtd_location_id = null,
+        $location_id = null, //do something with this
```
Is this comment still valid? It looks like it's used on line 78.

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
The complexity of the `run` method is pretty high. Can we pull out some of the stuff in `run` into named methods for clarity? I'll add details below.

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
```diff
@@ -0,0 +1,143 @@
+
+        if ($request->has('image')) {
+            $asset = $request->handleImages($asset);
+        }
```
This could be something like `private static function handleImages()`

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
```diff
@@ -0,0 +1,143 @@
+                    }
+                }
+            }
+        }
```
This could be something like `private static function handleCustomFields()`

marcusmoore commented on app/Actions/Assets/StoreAssetAction.php:
```diff
@@ -0,0 +1,143 @@
+
+        if (isset($target)) {
+            $asset->checkOut($target, auth()->user(), date('Y-m-d H:i:s'), $request->input('expected_checkin', null), 'Checked out on asset creation', $request->get('name'), $location);
+        }
```
This could be something like `private static function handleCheckout()`

marcusmoore commented on app/Actions/Assets/UpdateAssetAction.php:
```diff
@@ -0,0 +1,249 @@
+        $asset_tag = null,
+        $notes = null,
+        $isBulk = false,
+    ): \App\Models\SnipeModel {
```
Can we import this por favor

marcusmoore commented on app/Actions/Assets/UpdateAssetAction.php:
Same comment about improving readability of this complex method. I'll add some ideas below.

Thought: the chunk of `if`s isn't fun to read but it's _probably_ not worth extracting methods for them...

spencerrlongg replied:
those could probably do with a full refactor, but not something I think makes sense for this PR, definitely worth revisiting later.

marcusmoore commented on app/Actions/Assets/UpdateAssetAction.php:
```diff
@@ -0,0 +1,249 @@
+
+        if (isset($target)) {
+            $asset->checkOut($target, auth()->user(), date('Y-m-d H:i:s'), '', 'Checked out on asset update', e($request->get('name')), $location);
+        }
```
This could be something like `private static function handleCheckout()`

marcusmoore commented on app/Actions/Assets/UpdateAssetAction.php:
```diff
@@ -0,0 +1,249 @@
+            }
+        }
+
+        session()->put(['redirect_option' => $request->get('redirect_option'), 'checkout_to_type' => $request->get('checkout_to_type')]);
```
This feels like it should be handled at the controller level if possible.

marcusmoore commented on app/Actions/Assets/UpdateAssetAction.php:
```diff
@@ -0,0 +1,249 @@
+                    $asset->{$field->db_column} = $field_val;
+                }
+            }
+        }
```
Same comment about `handleCustomFields()`

marcusmoore commented on app/Exceptions/Handler.php:
```diff
@@ -2,6 +2,7 @@
 
 namespace App\Exceptions;
 
+use Illuminate\Database\Eloquent\ModelNotFoundException;
```
This can be removed.

marcusmoore commented on app/Http/Requests/Assets/UpdateAssetRequest.php:
```diff
@@ -28,11 +29,22 @@ public function authorize()
     public function rules()
     {
+        $modelRules = (new Asset)->getRules();
+        // TODO: make sure this actually works
```
Has this been tested? Can we remove this todo?

marcusmoore commented on tests/Feature/Assets/Api/UpdateAssetTest.php:
```diff
@@ -86,7 +86,7 @@ public function testAllAssetAttributesAreStored()
         $this->assertEquals($userAssigned->id, $updatedAsset->assigned_to);
         $this->assertTrue($updatedAsset->company->is($company));
-        $this->assertTrue($updatedAsset->location->is($location));
+        $this->assertTrue($updatedAsset->location->is($location)); //fix all location setting
```
What's this comment mean?

marcusmoore commented on tests/Feature/Assets/Ui/BulkEditAssetsTest.php:
```diff
@@ -221,11 +221,12 @@ public function testBulkEditAssetsAcceptsAndUpdatesEncryptedCustomFields()
         $this->markIncompleteIfMySQL('Custom Fields tests do not work on mysql');
         $edit_user = User::factory()->editAssets()->create();
-        $admin_user = User::factory()->admin()->create();
+        // admin used to work, but now only superuser does????
```
Is this comment still accurate? Maybe we should investigate...

spencerrlongg replied:
added the correct permission to the admin factory and reverted

marcusmoore commented on tests/Feature/Assets/Ui/StoreAssetTest.php:
```diff
@@ -0,0 +1,149 @@
+use Illuminate\Http\UploadedFile;
+use Illuminate\Support\Facades\Storage;
+use Tests\TestCase;
+use function Amp\Promise\wait;
```
This import can be removed.

marcusmoore commented on tests/Unit/DepreciationTest.php:
```diff
@@ -38,8 +38,9 @@ public function testDepreciationAmount()
-                    'category_id' => Category::factory()->assetLaptopCategory()->create(),
-                    'purchase_date' => now()->subDecade(),
+                    //not sure how this ever worked... do these need a category??
+                    //'category_id' => Category::factory()->assetLaptopCategory()->create(),
```
I think we can remove this comment. I think turning on validation exceptions is causing this?

marcusmoore commented on tests/Unit/DepreciationTest.php:
```diff
@@ -62,8 +63,9 @@ public function testDepreciationPercentage()
-                    'category_id' => Category::factory()->assetLaptopCategory()->create(),
-                    'purchase_date' => now()->subDecade(),
+                    //not sure how this ever worked...
+                    //'category_id' => Category::factory()->assetLaptopCategory()->create(),
```
I think we can remove this comment. I think turning on validation exceptions is causing this?

marcusmoore commented on app/Http/Controllers/Api/AssetsController.php:
```diff
@@ -602,87 +603,43 @@ public function selectlist(Request $request): array
+        } catch (CheckoutNotAllowed $e) {
+            return response()->json(Helper::formatStandardApiResponse('error', null, $e->getMessage()), 200);
+        } catch (Exception $e) {
+            return response()->json(Helper::formatStandardApiResponse('error', null, $e->getMessage()));
```
Do we want to add a `report()` here?

marcusmoore commented on app/Http/Controllers/Assets/BulkAssetsController.php:
```diff
@@ -2,9 +2,13 @@
 
 namespace App\Http\Controllers\Assets;
 
+use App\Actions\Assets\StoreAssetAction;
```
This line can be removed

marcusmoore commented on app/Http/Controllers/Assets/BulkAssetsController.php:
```diff
@@ -199,298 +204,61 @@ public function edit(Request $request) : View | RedirectResponse
+            } catch (ValidationException $e) {
+                $errors = $e->validator->errors()->toArray();
+            } catch (CustomFieldPermissionException $e) {
+                //$errors[$key] = $e->getMessage();
```
This commented code can be removed

marcusmoore reviewed (changes requested):
😮‍💨 that was a doozy. Thanks for all of your work on this.

I haven't manually tried the code paths yet but I have (a bunch) of preliminary comments. Most of them are about removing commented code (keeping it cleaner so we don't carry it around forever) but some are legit questions about potential changes.

Thanks again for the effort. I know it's a lot.

spencerrlongg commented:
Awesome, thank you so much for all of that @marcusmoore - a lot of that I probably should've found/fixed myself, I think I've just been too close to it for a while. 

I just went through them and am with you on all of it, I'll implement it all in the morning.

spencerrlongg committed:
Refactor asset handling and streamline code

This refactoring introduces helper methods for handling image uploads, custom fields, and checkouts, improving code organization and readability. Unnecessary imports, comments, and unused variables were removed for cleaner code. Error handling was also improved with additional exception reporting in the asset update flow.

spencerrlongg committed:
rm a couple of comments

spencerrlongg committed:
Merge branch 'develop' into refactor/asset-action

spencerrlongg committed:
added a permission to admin factory

spencerrlongg committed:
Merge remote-tracking branch 'origin/refactor/asset-action' into refactor/asset-action

spencerrlongg committed:
Merge branch 'develop' into refactor/asset-action

spencerrlongg committed:
fix test

spencerrlongg committed:
derp, that took way too long
